consensus/src/lib.rs

```diff
@@ -50,11 +63,101 @@ pub enum ConsensusCreationError {
     MiningStopped,
     #[error("Mining failed")]
     MiningFailed,
+    #[error("Staking error")]
+    StakingError(#[from] ConsensusPoSError),
+    #[error("Staking failed")]
+    StakingFailed,
+    #[error("Staking stopped")]
+    StakingStopped,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub enum GenerateBlockInputData {
+    PoW,
+    PoS(PoSGenerateBlockInputData),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct PoSGenerateBlockInputData {
+    vrf_private_key: VRFPrivateKey,
+    pool_id: PoolId,
+    kernel_input: TxInput,
+    kernel_witness: InputWitness,
+}
+
+impl PoSGenerateBlockInputData {
+    pub fn vrf_private_key(&self) -> &VRFPrivateKey {
+        &self.vrf_private_key
+    }
+
+    pub fn pool_id(&self) -> PoolId {
+        self.pool_id
+    }
+
+    pub fn kernel_input(&self) -> &TxInput {
+        &self.kernel_input
+    }
+
+    pub fn kernel_witness(&self) -> &InputWitness {
+        &self.kernel_witness
+    }
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub enum FinalizeBlockInputData {
+    PoW,
+    PoS(PoSFinalizeBlockInputData),
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct PoSFinalizeBlockInputData {
+    vrf_private_key: VRFPrivateKey,
+    epoch_index: EpochIndex,
+    sealed_epoch_randomness: PoSRandomness,
+    pool_balance: Amount,
```

Comment:
Why not pull the balance from the database, from accounting?

Reply:
Yep, that's what I'll be doing. I just left this as one of the last TODO items rather than being any issue